Add `AssetAndTimeSchedule` timetable

aaron-y-chen opened the pull request:
related: #58056

# Why

Use case: #58056

# How

### 1. Add `AssetAndTimeSchedule`

Add a timetable that combines a time-based timetable with an asset condition.

It behaves like a scheduled timetable, not an asset-triggered timetable: when it creates a run, the run type is still `SCHEDULED`.

### 2. Gate DagRun creation before creating the DagRun

The scheduler creates a scheduled DagRun only when both conditions are satisfied: the schedule time is due and the required asset condition is ready.

If assets are missing, no DagRun is created and `next_dagrun_create_after` stays on the pending slot, so the same slot can be retried later without creating placeholder DagRuns.

### 3. Keep asset-aware scheduler behavior consistent

Timetables expose asset-aware scheduling through `asset_triggered` and `asset_gated` behavior flags, following the existing `periodic` pattern. These flags are also mirrored on the Task SDK base timetable so custom timetables can opt in without scheduler type checks.

`DagModel.dags_needing_dagruns` uses these flags to route asset-triggered Dags separately from asset-gated scheduled Dags. Asset-gated Dags still flow through the standard scheduled-run creation path (`_create_dag_runs`).

When `DagModel.timetable_asset_gated` is set, the scheduler re-checks the timetable's `asset_condition` under ADRQ row locks before creating the scheduled DagRun. If the condition is still satisfied, the created run links and consumes the selected asset events. The ADRQ locking, event-provenance, and consumption helpers are shared with the asset-triggered path.

# What

Here are my test DAGs:

`upstream_asset_producer`
```python
import pendulum
from airflow.models.dag import DAG
from airflow.operators.bash import BashOperator
from airflow.sdk.definitions.asset import Asset

my_asset = Asset("/my/example/asset")

with DAG(
    dag_id="upstream_asset_producer",
    start_date=pendulum.datetime(2025, 1, 1, tz="UTC"),
    schedule="0 * * * *",
    catchup=False,
    tags=["example", "upstream"],
) as upstream_dag:
    BashOperator(
        task_id="produce_asset",
        bash_command="echo 'Asset produced!'",
        outlets=[my_asset],
    )
```

`downstream_asset_and_time_consumer`
```python
import pendulum
from airflow.models.dag import DAG
from airflow.operators.bash import BashOperator
from airflow.sdk.definitions.asset import Asset
from airflow.timetables.trigger import CronTriggerTimetable
from airflow.timetables.assets import AssetAndTimeSchedule

my_asset = Asset("/my/example/asset")

with DAG(
    dag_id="downstream_asset_and_time_consumer",
    start_date=pendulum.datetime(2025, 1, 1, tz="UTC"),
    schedule=AssetAndTimeSchedule(
        timetable=CronTriggerTimetable("4 * * * *", timezone="UTC"),
        assets=[my_asset],
    ),
    catchup=False,
    max_active_runs=1,
    tags=["example", "downstream", "asset-and-time"],
) as downstream_dag:
    BashOperator(
        task_id="consume_asset",
        bash_command="echo 'Asset condition and time condition were both met!'",
    )
```

From the screenshots, the downstream DAG runs as expected: it runs only after both the timetable and the assets are ready.

1. Upstream dag  

   ![AssetAndTimeSchedule_example_1_upstream](https://github.com/user-attachments/assets/ed205b0e-78eb-453d-a83b-26bd6e29f42e)

2. Asset production time  

   ![AssetAndTimeSchedule_example_2_asset](https://github.com/user-attachments/assets/3224c304-ffd2-4f60-a24e-5431692d6142)

3. Downstream, `AssetAndTimeSchedule`  

   ![AssetAndTimeSchedule_example_3_downstream](https://github.com/user-attachments/assets/dbd840ef-8059-47d3-a9af-dfa27b8227ef)

<!-- pr-triage-fold: triaged=2026-06-22T06:31:40.647663+00:00 head=6a7695a action=comment -->

---

> [!IMPORTANT]
> **🛠️ Maintainer triage note for @nailo2c** · by `@potiuk` · 2026-06-22 06:31 UTC
>
> Your review threads from `@jscheffl` look addressed — please confirm this PR is **ready for maintainer review confirmation**:
> - 1 thread(s) show your engagement (post-review commits and/or replies).
>
> **The ball is in your court** — you've been assigned to this PR. Reply `yes / ready` (and mark the threads resolved) and a maintainer will pick it up from the queue.
>
> <sub>_Automated triage — may be imperfect; a maintainer takes the next look._</sub>

<!-- /pr-triage-fold -->

nailo2c committed:
feat-58056: Implement AssetAndTimeSchedule for time-based scheduling with asset conditions

Lee-W commented on airflow-core/src/airflow/timetables/assets.py:
```diff
@@ -90,3 +94,75 @@ def generate_run_id(self, *, run_type: DagRunType, **kwargs: typing.Any) -> str:
         return super().generate_run_id(run_type=run_type, **kwargs)
+
+
+class AssetAndTimeSchedule(Timetable):
```
I think we'll also need a task sdk counter part

aaron-y-chen replied:
Really great suggestion, I saw that `AssetOrTimeSchedule` have a task sdk version, and yes this should have one too.

Lee-W commented on airflow-core/src/airflow/jobs/scheduler_job_runner.py:
```diff
@@ -2227,6 +2228,50 @@ def _update_state(dag: SerializedDAG, dag_run: DagRun):
                     continue
+            # For AssetAndTimeSchedule, defer starting until all required assets are queued.
+            if isinstance(dag.timetable, AssetAndTimeSchedule):
+                # Count required assets for this DAG's schedule
```
```suggestion
                # Count required assets for this Dag's schedule
```

Lee-W commented on airflow-core/src/airflow/jobs/scheduler_job_runner.py:
```diff
@@ -2227,6 +2228,50 @@ def _update_state(dag: SerializedDAG, dag_run: DagRun):
+            # For AssetAndTimeSchedule, defer starting until all required assets are queued.
+            if isinstance(dag.timetable, AssetAndTimeSchedule):
+                # Count required assets for this DAG's schedule
+                required_count = (
```
Instead of doing this, we probably should do something similar to `dag_ready` in `dag.py`

aaron-y-chen replied:
Good point, `AssetEvaluator` was introduced after this feature was originally implemented, and it looks applicable here as well. Let me refactor this part to use it.

Lee-W commented on airflow-core/src/airflow/models/dag.py:
```diff
@@ -679,6 +679,12 @@ def dag_ready(dag_id: str, cond: SerializedAssetBase, statuses: dict[UKey, bool]
         for ser_dag in ser_dags:
             dag_id = ser_dag.dag_id
             statuses = dag_statuses[dag_id]
+            timetable = ser_dag.dag.timetable
```
What is this part for?

aaron-y-chen replied:
This part is to guarantee that only `AssetTriggeredTimetable` can be added to `triggered_date_by_dag`, please check the "Discussion" part in the PR description.

nailo2c committed:
Apply suggestion from @Lee-W

Co-authored-by: Wei Lee <weilee.rx@gmail.com>

nailo2c committed:
refactor - using AssetEvaluator & add task sdk interface

nailo2c committed:
Handle asset wait timeout for AssetAndTimeSchedule

nailo2c committed:
add doc for task sdk

nailo2c committed:
fix CI error

nailo2c committed:
fix CI error #2

Lee-W commented on airflow-core/docs/authoring-and-scheduling/asset-scheduling.rst:
```diff
@@ -147,6 +147,30 @@ If one asset is updated multiple times before all consumed assets update, the do
+
+Use ``AssetAndTimeSchedule`` when you want a Dag to follow a normal time-based timetable but only start after specific assets have been updated. Airflow creates the DagRun at the scheduled time and keeps it queued until every required asset has queued an event. When the DagRun starts, those asset events are consumed so the next scheduled run waits for new updates. This does not create additional  ...
+
+If ``dagrun_timeout`` is set on the DAG, it also limits how long an ``AssetAndTimeSchedule`` run may remain queued while waiting for its asset condition to become ready.
```
```suggestion
If ``dagrun_timeout`` is set on the Dag, it also limits how long an ``AssetAndTimeSchedule`` run may remain queued while waiting for its asset condition to become ready.
```

Lee-W commented on task-sdk/src/airflow/sdk/__init__.pyi:
```diff
@@ -107,6 +108,7 @@ __all__ = [
     "AssetAlias",
     "AssetAll",
     "AssetAny",
+    "AssetAndTimeSchedule",
```
```suggestion
    "AssetAndTimeSchedule",
    "AssetAny",
```

We sort it in alphabetical order

aaron-y-chen replied:
Nice catch, I'll keep that in mind

Lee-W commented on task-sdk/src/airflow/sdk/__init__.py:
```diff
@@ -169,6 +171,7 @@
     "AssetAlias": ".definitions.asset",
     "AssetAll": ".definitions.asset",
     "AssetAny": ".definitions.asset",
+    "AssetAndTimeSchedule": ".definitions.timetables.assets",
```
```suggestion
    "AssetAndTimeSchedule": ".definitions.timetables.assets",
    "AssetAny": ".definitions.asset",
```

aaron-y-chen replied:
Nice catch

Lee-W reviewed (commented):
1. Where exactly is the logic for "2. Assign AssetAndTimeSchedule to the non_asset_dags set" added?
2. Can you explain the purpose of the timeout logic? For example, if a Dag is scheduled for 17:00, does it wait until 17:00 + timeout? If the asset events do not arrive, the task fails — what happens to the existing ADRQs? Are they cleared, or do they count toward the next schedule?
3. What will the UI look like in each waiting state? Specifically:
   - If asset events have arrived but the scheduled time hasn't come yet, does the UI indicate we're waiting for the scheduled time?
   - If the scheduled time has passed but the asset event hasn't arrived yet, does the UI show which assets are still being waited on?

nailo2c committed:
Update task-sdk/src/airflow/sdk/__init__.py

Co-authored-by: Wei Lee <weilee.rx@gmail.com>

nailo2c committed:
Update task-sdk/src/airflow/sdk/__init__.pyi

Co-authored-by: Wei Lee <weilee.rx@gmail.com>

nailo2c committed:
Term modify: DAG -> Dag

Co-authored-by: Wei Lee <weilee.rx@gmail.com>

nailo2c committed:
Enhance AssetAndTimeSchedule handling by deferring scheduled runs until all required assets are queued

Lee-W commented on airflow-core/src/airflow/jobs/scheduler_job_runner.py:
```diff
@@ -2264,6 +2266,63 @@ def _update_state(dag: SerializedDAG, dag_run: DagRun):
+                            DagRunType.SCHEDULED,
+                            DagRunType.MANUAL,
+                            DagRunType.ASSET_TRIGGERED,
+                        )
```
Do we need to check this again? We've checked it in line 2271

aaron-y-chen replied:
Ah... yes that's redundant, since line 2271 has set DagRunType to SCHEDULED, we don't need to check it here again. Great catch.

Lee-W commented on airflow-core/src/airflow/jobs/scheduler_job_runner.py:
```diff
@@ -2264,6 +2266,63 @@ def _update_state(dag: SerializedDAG, dag_run: DagRun):
+                        )
+                        and dag_run.dag_model is not None
+                    ):
+                        self._set_exceeds_max_active_runs(dag_model=dag_run.dag_model, session=session)
```
it's a failed run now. do we need to check max_active_runs here

aaron-y-chen replied:
I think it's still necessary. Once the queued run fails, we need to update `dag_model.exceeds_max_non_backfill`. Otherwise Scheduler may still think the Dag is maxed out and skip creating the next scheduled run.

https://github.com/apache/airflow/blob/b3138e2400eeed96863c9ced3ff2ff3e0ad7a519/airflow-core/src/airflow/jobs/scheduler_job_runner.py#L3314

Lee-W reviewed (commented):
mostly good, 2 questions

nailo2c committed:
Fix redundant DagRun type check

Lee-W commented on airflow-core/src/airflow/timetables/assets.py:
```diff
@@ -90,3 +94,75 @@ def generate_run_id(self, *, run_type: DagRunType, **kwargs: typing.Any) -> str:
+        self,
+        *,
+        timetable: Timetable,
+        assets: Collection[Asset] | BaseAsset | SerializedAssetBase,
```
```suggestion
        assets: Collection[SerializedAsset] | SerializedAssetBase,
```

I think we should be able to use it now?

aaron-y-chen replied:
Also a nice catch, this should migrate to `SerializedAsset`

Lee-W reviewed (commented):
one last question, but mostly good

nailo2c committed:
Update airflow-core/src/airflow/timetables/assets.py

Co-authored-by: Wei Lee <weilee.rx@gmail.com>

nailo2c committed:
Remove unused Asset import and fix test fixtures for new annotation

The annotation on AssetAndTimeSchedule.__init__ was migrated to
Collection[SerializedAsset] | SerializedAssetBase, leaving the
TYPE_CHECKING-only Asset import unreferenced and breaking mypy on
two test files that passed SDK Assets to the core class. The
scheduler tests now import AssetAndTimeSchedule (and the matching
CronTriggerTimetable) from airflow.sdk so the types align with how
DAG authors use it; the timetable test fixture mirrors
core_asset_timetable by constructing SerializedAsset directly.

jscheffl commented on airflow-core/src/airflow/jobs/scheduler_job_runner.py:
```diff
@@ -2582,6 +2584,55 @@ def _update_state(dag: SerializedDAG, dag_run: DagRun):
                     continue
+            # For AssetAndTimeSchedule, defer starting until all required assets are queued.
+            # Only gate scheduled runs; manual and backfill runs should start immediately.
+            if isinstance(dag.timetable, AssetAndTimeSchedule) and dag_run.run_type == DagRunType.SCHEDULED:
```
I think I understand the demand, but not being an expert in timetables and scheduler... looking at the code I am not fully convinced that the proposed implementation strategy is "good". It assumes the Dag is triggered to queued state at point of schedule and then is sitting "queued" waiting for events. I think this is the wrong state model. In my view the trigger should only happen at point of schedule if events are available. In all other cases if should not generate a Dag run in my view and this also should not mark the Dag as failed just because of missing events.

Marking a queued Dag run to failed is the wrong signal, feels like something is broken. Also besides this the code you are using is not updating the tasks to be skipped but they stay in the state like before ("none"), usually if a Dag run is set to failed, tasks are set to skipped.

aaron-y-chen replied:
Thanks Jens, this is a really, really helpful insight. I'm checking the fundamental mechanism of ADRQ and I found there might be a clean way to design for this case.

Please give me a couple days to study and think, and I'll try to propose a better design later.

aaron-y-chen replied:
+ Old approach: when the schedule time was reached, the scheduler created a placeholder `QUEUED` DagRun first, then waited for asset events in a later scheduler loop.

+ New approach: the scheduler creates a scheduled DagRun only when both conditions are satisfied: the schedule time is due and the required asset condition is ready. If assets are missing, no DagRun is created and `next_dagrun_create_after` stays on the pending slot, so the same slot can be retried later without placeholder DagRuns.

I feel the new approach is much cleaner and without hack to handle the timeout issue, and the gating point is moved from after DagRun creation to before DagRun creation.

aaron-y-chen committed:
Add AssetAndTimeSchedule timetable for time + asset gated scheduling

Add a new built-in timetable that combines a time-based schedule with
an asset condition: a SCHEDULED DagRun is created only when both the
timetable's scheduled time has arrived and every required asset has
queued an event. When the run is created, those asset events are
consumed so the next scheduled run waits for fresh updates. Unlike
AssetOrTimeSchedule, this does not create asset-triggered runs.

Implementation outline:
- airflow.timetables.assets.AssetAndTimeSchedule (core) and
  airflow.sdk.definitions.timetables.assets.AssetAndTimeSchedule
  (Task SDK shim) delegate next_dagrun_info / infer_manual_data_interval
  / generate_run_id to the wrapped timetable, retaining the
  asset_condition that gates creation.
- DagModel.dags_needing_dagruns returns a third bucket
  asset_gated_ready_dag_ids of dags whose asset condition is already
  satisfied; the SQL query excludes AssetAndTimeSchedule dags whose
  assets are not yet ready, preventing them from saturating the
  max_dagruns_to_create_per_loop batch and starving pure time dags.
- _create_dagruns_for_dags dispatches the new bucket to
  _create_dag_runs_asset_gated. The method re-locks ADRQ rows with
  skip_locked, re-evaluates the asset condition as a HA race guard,
  creates a SCHEDULED DagRun, links the consumed AssetEvent rows via
  dag_run.consumed_asset_events.extend(...), updates active_runs and
  exceeds_max_non_backfill so a follow-up asset event cannot bypass
  max_active_runs, and finally deletes only the consumed ADRQ rows.
- _start_queued_dagruns is untouched: asset gating happens at DagRun
  creation time, not after. There is no placeholder QUEUED DagRun
  waiting for assets and no dagrun_timeout-based FAILED path, both of
  which would have violated the DagRun state-model invariants
  documented in airflow-core/docs/core-concepts/dag-run.rst.

Tests:
- AssetAndTimeSchedule serialization/deserialization, data-interval
  inference, next_dagrun_info delegation, and generate_run_id.
- Scheduler-level: time-not-ready, assets-not-ready, both-ready,
  oldest-pending-slot semantics on late asset arrival, only-consumes-one-slot
  when multiple schedule slots are pending, ADRQ row locking race guard,
  max_active_runs respected after a queued run is created, AssetEvent
  rows linked via consumed_asset_events, and starvation prevention
  showing the new SQL filter keeps gated dags from starving pure time
  dags when assets are missing.
- DagModel.dags_needing_dagruns regression coverage for the new
  third-bucket return value.

closes: #58056

aaron-y-chen committed:
Add AssetAndTimeSchedule timetable for time + asset gated scheduling

Add a new built-in timetable that combines a time-based schedule with
an asset condition: a SCHEDULED DagRun is created only when both the
timetable's scheduled time has arrived and every required asset has
queued an event. When the run is created, those asset events are
consumed so the next scheduled run waits for fresh updates. Unlike
AssetOrTimeSchedule, this does not create asset-triggered runs.

Implementation outline:
- airflow.timetables.assets.AssetAndTimeSchedule (core) and
  airflow.sdk.definitions.timetables.assets.AssetAndTimeSchedule
  (Task SDK shim) delegate next_dagrun_info / infer_manual_data_interval
  / generate_run_id to the wrapped timetable, retaining the
  asset_condition that gates creation.
- DagModel.dags_needing_dagruns returns a third bucket
  asset_gated_ready_dag_ids of dags whose asset condition is already
  satisfied; the SQL query excludes AssetAndTimeSchedule dags whose
  assets are not yet ready, preventing them from saturating the
  max_dagruns_to_create_per_loop batch and starving pure time dags.
- _create_dagruns_for_dags dispatches the new bucket to
  _create_dag_runs_asset_gated. The method re-locks ADRQ rows with
  skip_locked, re-evaluates the asset condition as a HA race guard,
  creates a SCHEDULED DagRun, links the consumed AssetEvent rows via
  dag_run.consumed_asset_events.extend(...), updates active_runs and
  exceeds_max_non_backfill so a follow-up asset event cannot bypass
  max_active_runs, and finally deletes only the consumed ADRQ rows.
- _start_queued_dagruns is untouched: asset gating happens at DagRun
  creation time, not after. There is no placeholder QUEUED DagRun
  waiting for assets and no dagrun_timeout-based FAILED path, both of
  which would have violated the DagRun state-model invariants
  documented in airflow-core/docs/core-concepts/dag-run.rst.

Tests:
- AssetAndTimeSchedule serialization/deserialization, data-interval
  inference, next_dagrun_info delegation, and generate_run_id.
- Scheduler-level: time-not-ready, assets-not-ready, both-ready,
  oldest-pending-slot semantics on late asset arrival, only-consumes-one-slot
  when multiple schedule slots are pending, ADRQ row locking race guard,
  max_active_runs respected after a queued run is created, AssetEvent
  rows linked via consumed_asset_events, and starvation prevention
  showing the new SQL filter keeps gated dags from starving pure time
  dags when assets are missing.
- DagModel.dags_needing_dagruns regression coverage for the new
  third-bucket return value.

closes: #58056

aaron-y-chen committed:
Add docs for AssetAndTimeSchedule

Document the new gating model: a scheduled DagRun is created only
when both the timetable's scheduled time has arrived and every
required asset has queued an event. If the scheduled time arrives
before assets are ready, no DagRun is created; the scheduler holds
the oldest pending slot and re-checks each loop until assets arrive.

closes: #58056

aaron-y-chen committed:
Add docs for AssetAndTimeSchedule

Document the new gating model: a scheduled DagRun is created only
when both the timetable's scheduled time has arrived and every
required asset has queued an event. If the scheduled time arrives
before assets are ready, no DagRun is created; the scheduler holds
the oldest pending slot and re-checks each loop until assets arrive.

closes: #58056

aaron-y-chen committed:
Refactor asset gated DAG run creation logic for clarity and efficiency

aaron-y-chen committed:
Refactor asset gated DAG run creation logic for clarity and efficiency

aaron-y-chen committed:
Use timetable behavior flags for asset-aware scheduling

aaron-y-chen committed:
Use timetable behavior flags for asset-aware scheduling

uranusjr commented on airflow-core/src/airflow/jobs/scheduler_job_runner.py:
```diff
@@ -2768,6 +2770,55 @@ def _update_state(dag: SerializedDAG, dag_run: DagRun):
                     continue
+            # For AssetAndTimeSchedule, defer starting until all required assets are queued.
+            # Only gate scheduled runs; manual and backfill runs should start immediately.
+            if isinstance(dag.timetable, AssetAndTimeSchedule) and dag_run.run_type == DagRunType.SCHEDULED:
```
I don’t think it is a good idea to handle timetables by special-casing the type like this. Timetables are designed to provide encapsulation around scheduling to begin with, and this kind of code structuring defeats the purpose. Please think about implementation at a higher level to make the interface more modular, similar to how various timetables implement time-based scheduling differently, and how asset partitioning uses additional abstraction to handle assets differently agasint different configurations.

aaron-y-chen replied:
I found that while resolving conflicts a few days ago, I accidentally force-pushed from a stale checkout, so the code you reviewed was not the behavioral flag version I mentioned... I'm really, really sorry about that.

I have restored the latest version now. I'll be more careful going forward and make sure this doesn't happen again. 😫

aaron-y-chen committed:
Route asset-gated Dags through the standard scheduled-run path

Review feedback asked for the scheduler to handle timetables through
behavioral contracts instead of special-cased code paths. The dedicated
asset-gated creation path duplicated both the scheduled-run creation
logic and the asset-triggered ADRQ/event mechanics, and could drift
from either. Asset gating is now a generic step of the normal
scheduled-run path, driven only by DagModel.timetable_asset_gated and
Timetable.asset_condition, with the ADRQ locking, event-provenance and
consumption helpers shared with the asset-triggered path.

aaron-y-chen committed:
Route asset-gated Dags through the standard scheduled-run path

Review feedback asked for the scheduler to handle timetables through
behavioral contracts instead of special-cased code paths. The dedicated
asset-gated creation path duplicated both the scheduled-run creation
logic and the asset-triggered ADRQ/event mechanics, and could drift
from either. Asset gating is now a generic step of the normal
scheduled-run path, driven only by DagModel.timetable_asset_gated and
Timetable.asset_condition, with the ADRQ locking, event-provenance and
consumption helpers shared with the asset-triggered path.

aaron-y-chen committed:
Add regression test to prevent reattribution of consumed asset events in gated runs

aaron-y-chen committed:
Add regression test to prevent reattribution of consumed asset events in gated runs

shahar1 commented on airflow-core/src/airflow/timetables/assets.py:
```diff
@@ -90,3 +93,77 @@ def generate_run_id(self, *, run_type: DagRunType, **kwargs: typing.Any) -> str:
+        if self.timetable.asset_triggered or self.timetable.asset_gated:
+            raise AirflowTimetableInvalid("cannot nest asset-aware timetables")
+        if not isinstance(self.asset_condition, SerializedAssetBase):
+            raise AirflowTimetableInvalid("all elements in 'assets' must be assets")
```
Duplicates `AssetOrTimeSchedule.validate()`, maybe we could use a helper function?

aaron-y-chen replied:
Good idea, let me do it later :)

shahar1 commented on airflow-core/src/airflow/jobs/scheduler_job_runner.py:
```diff
@@ -2551,6 +2568,128 @@ def _create_dag_runs(self, dag_models: Collection[DagModel], session: Session) -
+                    # A gated run consumes events newer than its run_after (the slot
+                    # time), so the previous-run floor alone would re-attribute them to
+                    # the next run; skip events already consumed by this Dag's runs.
+                    ~AssetEvent.created_dagruns.any(DagRun.dag_id == dag_id),
```
Redundant on the asset-triggered caller - there a run's `run_after` equals the max consumed-event timestamp, so `timestamp > floor` above already excludes prior-consumed events; only the gated path needs this.

aaron-y-chen replied:
ah, good catch, let me fix it.

aaron-y-chen committed:
Enhance asset scheduling by adding event exclusion logic and refactoring validation checks

aaron-y-chen committed:
Enhance asset scheduling by adding event exclusion logic and refactoring validation checks

uranusjr commented on task-sdk/src/airflow/sdk/definitions/timetables/assets.py:
```diff
@@ -84,3 +85,23 @@ class AssetOrTimeSchedule(AssetTriggeredTimetable):
+    asset_condition: BaseAsset = attrs.field(alias="assets", converter=_coerce_assets)
+    timetable: BaseTimetable
+
+    def __attrs_post_init__(self) -> None:
+        self.active_runs_limit = self.timetable.active_runs_limit
+        self.can_be_scheduled = self.timetable.can_be_scheduled
```
Should these be properties instead?

aaron-y-chen replied:
If I change it to `properties`, do I also need to update the `AssetOrTimeSchedule` part?

https://github.com/apache/airflow/blob/e24f5294645c20f55069c2f2a1b890eb3ff8cc2a/task-sdk/src/airflow/sdk/definitions/timetables/assets.py#L84-L86

uranusjr commented on airflow-core/src/airflow/timetables/assets.py:
```diff
@@ -20,7 +20,10 @@
-from airflow.serialization.definitions.assets import SerializedAsset, SerializedAssetBase
+from airflow.sdk.definitions.asset import BaseAsset  # TODO: Use serialized classes.
+from airflow.serialization.definitions.assets import SerializedAsset, SerializedAssetAll, SerializedAssetBase
+from airflow.serialization.encoders import ensure_serialized_asset
```
This should be imported inside the encode function (similar to other timetables). Encoders are heavy, and a serialized timetable does not need to be re-encoded in most cases.

aaron-y-chen replied:
Good idea, let me do it.

uranusjr commented on airflow-core/src/airflow/timetables/assets.py:
```diff
@@ -20,7 +20,10 @@
 
 from airflow.exceptions import AirflowTimetableInvalid
-from airflow.serialization.definitions.assets import SerializedAsset, SerializedAssetBase
+from airflow.sdk.definitions.asset import BaseAsset  # TODO: Use serialized classes.
```
This is a bad idea. I’d swap the isinstance check below to act on serialized objects, and default all unrecognized types to `ensure_serialized_asset` (and let that function blow up) instead.

aaron-y-chen replied:
Fixed. I've already removed it. Thanks!

aaron-y-chen committed:
Keep core AssetAndTimeSchedule free of SDK asset classes

aaron-y-chen committed:
Keep core AssetAndTimeSchedule free of SDK asset classes

aaron-y-chen committed:
Keep AssetAndTimeSchedule compatible with latest migrations

The PR and main added separate migrations from the same revision, so the branch must follow main's new head before it can merge cleanly.

aaron-y-chen committed:
Refactor AssetAndTimeSchedule properties for active_runs_limit and can_be_scheduled
